### PR TITLE
Add PL Trainer

### PR DESCRIPTION
**Pull request by boat1603**
## Why this PR
Add PL Trainer for LLM

## Changes
- Add PL Trainer scripts.

## Related Issues
Close #

## Checklist
- [ ] PR should be in the [Naming convention](../../docs/PR_NAMING.md)
- [ ] Assign yourself in to Assigneees
- [ ] Tag related issues
- [ ] Constants name should be ALL_CAPITAL, function name should be snake_case, and class name should be CamelCase
- [ ] complex function/algorithm should have [Docstring](https://peps.python.org/pep-0257/)
- [ ] 1 PR should not have more than 200 lines changes (Exception for test files). If more than that please open multiple PRs
- [ ] At least PR reviewer must come from the task's team (model, eval, data)

### CODE REVIEW

File: src/model/openthaigpt_pretraining_model/pl_trainer/optimizer/__init__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+
+def get_optimizer(parameters, optimizer_config):
+    optimizer_module = OPTIMIZERS.get(optimizer_config.name, None)
+    if optimizer_module is None:
+        raise ValueError(f"No optimizer name `{optimizer_config.name}`")
+    return optimizer_module(parameters, **optimizer_config.hyps)
+
+
+def get_scheduler(optimizer, base_lr, total_steps, scheduler_config):
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/pl_trainer/optimizer/__init__.py:11]: Can we have type hint for this function?

File: src/model/openthaigpt_pretraining_model/pl_trainer/optimizer/scheduler.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
+# Implement from https://github.com/mlfoundations/open_clip/blob/main/src/training/scheduler.py
+import numpy as np
+import torch
+import math
+
+INITIAL_LR = "initial_lr"
+
+
+def _warmup_lr(base_lr, warmup_length, step):
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/pl_trainer/optimizer/scheduler.py:9]: Can we have type hint for this function?

File: src/model/pyproject.toml [hunk 3]
```diff
@@ -25,11 +25,10 @@ dependencies = [
...
     "lightning>=2.0.2",
-    "bitsandbytes-cuda110",
     "wandb>=0.15.2",
     "bitsandbytes",
     "deepspeed>=0.9.2",
-    "nlpo3>=1.3.0",
     "peft>=0.3.0",
-    "scipy<2.0.0"
+    "scipy<2.0.0",
+    "tensorboard",
```
Comment (new5558) [inline src/model/pyproject.toml:33]: Version?

Comment (new5558) [reply src/model/pyproject.toml:33]: We can use 
- `^x.x.x` = update to latest version of `x.x.x` except major version change
- `~x.x.x` = update to latest version of `x.x.x` except major and minor version change

File: src/model/scripts/pytorch_lighting_training/pl_configurations/checkpointing/default_checkpointing.yaml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+save_dir: /path/to/save_dir
```
Comment (new5558) [inline src/model/scripts/pytorch_lighting_training/pl_configurations/checkpointing/default_checkpointing.yaml:1]: Can we make 2 files
- `default_checkpointing.example.yaml` - this file commits to git
- `default_checkpointing.yaml` - this file added in `.gitignore`
This will help prevent someone accidentally commit their new config file into our git

File: src/model/scripts/pytorch_lighting_training/pl_configurations/dataset/default_dataset.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+training_dataset: /path/to/train/data
```
Comment (new5558) [inline src/model/scripts/pytorch_lighting_training/pl_configurations/dataset/default_dataset.yaml:1]: Should we also do `example.yaml` thing here?

File: src/model/scripts/pytorch_lighting_training/pretraining.sh [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
```
Comment (new5558) [inline src/model/scripts/pytorch_lighting_training/pretraining.sh:1]: should we do `pretraining.example.sh` ?

File: src/model/scripts/pytorch_lighting_training/pl_configurations/config.yaml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+defaults:
+  name: llama-hf-7b-from-pretrained
+  - model: llama-hf-7b-from-pretrained
+  - optimizer: adamw_optimizer
+  - scheduler: warmup_and_decay_lr
+  - trainer: default_trainer
+  - deepspeed: stage_2
```
Comment (new5558) [inline src/model/scripts/pytorch_lighting_training/pl_configurations/config.yaml:7]: What should we do when there are other sharding methods e.g.. `FSDP`?

**Comment by new5558** [comment]
Can you fix black Linting and Formatting error krub. For pytest error we can ignore that for now.

File: src/model/pyproject.toml (hunk 3 above)
Comment (boat1603) [reply src/model/pyproject.toml:33]: I think we can use every version.

File: src/model/openthaigpt_pretraining_model/pl_trainer/optimizer/scheduler.py (hunk 2 above)
Comment (boat1603) [reply src/model/openthaigpt_pretraining_model/pl_trainer/optimizer/scheduler.py]: Done

File: src/model/scripts/pytorch_lighting_training/pl_configurations/dataset/default_dataset.yaml (hunk 5 above)
Comment (boat1603) [reply src/model/scripts/pytorch_lighting_training/pl_configurations/dataset/default_dataset.yaml]: Done

File: src/model/scripts/pytorch_lighting_training/pretraining.sh (hunk 6 above)
Comment (boat1603) [reply src/model/scripts/pytorch_lighting_training/pretraining.sh]: Done

File: src/model/scripts/pytorch_lighting_training/pl_configurations/config.yaml (hunk 7 above)
Comment (boat1603) [reply src/model/scripts/pytorch_lighting_training/pl_configurations/config.yaml]: I think we can handle these later.

File: src/model/pyproject.toml (hunk 3 above)
Comment (new5558) [reply src/model/pyproject.toml:33]: When a major version changes. code usually breaks

Comment (boat1603) [reply src/model/pyproject.toml]: Done

**Review by new5558 — APPROVED** [review]

**Review by ArthurMinovsky — APPROVED** [review]
